### ISSUE TITLE
Deploy fails sometimes due to race condition when starting netmon, vxlan-policy-agent and garden

### ISSUE THREAD

**Issue by giner** [cloudfoundry/cf-networking-release#65]
## Issue

When deploying Cells in parallel (max-in-flight is 8) on a big deployment (~500 cells) sometimes rep sometimes fails to start. This leads to a failure of the deployment process and the need to start it again. During the last deploy this happend ~10 times.

```
Task 30730 | 07:24:33 | Updating instance cell_stg2: cell_stg2/18d6f81b-6173-4824-947a-01f37f615168 (36) (00:15:27)
                     L Error: 'cell_stg2/18d6f81b-6173-4824-947a-01f37f615168 (36)' is not running after update. Review logs for failed jobs: rep
Task 30730 | 07:24:47 | Updating instance cell_stg2: cell_stg2/9f90ea46-d9e3-4633-9944-ae5474f96efc (18) (00:04:16)
Task 30730 | 07:25:51 | Updating instance cell_stg2: cell_stg2/a54fc6f0-43e1-4613-bcd1-b3ab958bc1ab (87) (00:05:04)
Task 30730 | 07:26:01 | Updating instance cell_stg2: cell_stg2/a8b88cb7-3b3f-4f20-b311-e57c35b9e369 (69) (00:05:04)
Task 30730 | 07:26:02 | Updating instance cell_stg2: cell_stg2/a611363f-20f6-45ef-af31-469b22c3f102 (24) (00:05:06)
Task 30730 | 07:26:08 | Updating instance cell_stg2: cell_stg2/a9d3b954-b06d-4331-aec2-563293ebb1d7 (66) (00:05:06)
Task 30730 | 07:26:10 | Updating instance cell_stg2: cell_stg2/9fcd8d5a-1ca8-4fc7-9e89-36adfc294156 (78) (00:05:23)
Task 30730 | 07:26:15 | Updating instance cell_stg2: cell_stg2/a993e227-75a5-4fd9-a0dd-3e328e5ed826 (57) (00:05:15)
Task 30730 | 07:26:15 | Error: 'cell_stg2/18d6f81b-6173-4824-947a-01f37f615168 (36)' is not running after update. Review logs for failed jobs: rep

Task 30730 Started  Fri Apr 19 03:47:47 UTC 2019
Task 30730 Finished Fri Apr 19 07:26:15 UTC 2019
Task 30730 Duration 03:38:28
Task 30730 error

Updating deployment:
  Expected task '30730' to succeed but state is 'error'

Exit code 1
```

## Context

cf-deployment v4.4.0
cf-networking 2.15.0

## Steps to Reproduce

1. Stop all jobs with `monit stop`
```
monit stop all
```
2. Remove /var/vcap/data/garden-cni
```
rm -r /var/vcap/data/garden-cni
```
3. Start garden, rep, consul_agent and loggregator_agent
```
monit start consul_agent
monit start loggregator_agent
monit start garden
monit start rep
```
4. Start the rest of jobs
```
monit start all
```

## Expected result

All jobs should start successfully

## Current result

rep is failing to start

Note the difference in permissions for `/var/vcap/data/garden-cni` directory. The problematic case doesn't have `x`. **The permissions depend on which job starts first and creates the folder** (race).

**Good**
```
# monit summary
The Monit daemon 5.2.5 uptime: 1h 49m 

Process 'consul_agent'              running
Process 'garden'                    running
Process 'rep'                       running
Process 'route_emitter'             running
Process 'netmon'                    running
Process 'vxlan-policy-agent'        running
Process 'silk-daemon'               running
Process 'loggregator_agent'         running
System 'system_localhost'           running

$ sudo ls -ld /var/vcap/data/garden-cni
drwx------ 3 vcap vcap 4096 Apr 18 04:37 /var/vcap/data/garden-cni

$ sudo su vcap -c "ls -l /var/vcap/data/garden-cni"
total 8
drw------- 2 root root 4096 Apr 18 04:37 container-netns
-rw------- 1 root root   22 Apr 18 04:37 external-networker-state.json
-rw------- 1 vcap vcap    0 Apr 18 04:37 iptables.lock
```

**Bad**
```
# monit summary
The Monit daemon 5.2.5 uptime: 1h 47m 

Process 'consul_agent'              running
Process 'garden'                    running
Process 'rep'                       Does not exist
Process 'route_emitter'             running
Process 'netmon'                    running
Process 'vxlan-policy-agent'        running
Process 'silk-daemon'               running
Process 'loggregator_agent'         running
System 'system_localhost'           running

$ sudo ls -ld /var/vcap/data/garden-cni
drw------- 3 vcap vcap 4096 Apr 18 04:27 /var/vcap/data/garden-cni

$ sudo su vcap -c "ls -l /var/vcap/data/garden-cni"
ls: cannot access /var/vcap/data/garden-cni/iptables.lock: Permission denied
ls: cannot access /var/vcap/data/garden-cni/external-networker-state.json: Permission denied
ls: cannot access /var/vcap/data/garden-cni/container-netns: Permission denied
total 0
d????????? ? ? ? ?            ? container-netns
-????????? ? ? ? ?            ? external-networker-state.json
-????????? ? ? ? ?            ? iptables.lock
```

rep.stdout.log:
```
...
{
  "timestamp": "2019-04-18T04:28:10.250161481Z",
  "level": "error",
  "source": "rep",
  "message": "rep.exited-with-failure",
  "data": {
    "error": "Exit trace for group:\ngarden_health_checker exited with error: external networker up: exit status 1\ncontainer-metrics-reporter exited with nil\nhub-closer exited with nil\nmetrics-reporter exited with nil\nvolman-driver-syncer exited with nil\ndebug-server exited with nil\n"
  }
}
...
```

vxlan-policy-agent.stdout.log:
```
...
{
  "timestamp": "2019-04-18T04:27:43.901602556Z",
  "level": "error",
  "source": "cfnetworking.vxlan-policy-agent",
  "message": "cfnetworking.vxlan-policy-agent.rules-enforcer.create-chain",
  "data": {
    "error": "lock: open lock file: open /var/vcap/data/garden-cni/iptables.lock: permission denied",
    "session": "4"
  }
}
...
```

## Possible Fix

**Quick fixes**
- Create a directory right after [removing](https://github.com/cloudfoundry/cf-networking-release/blob/v2.3.0/jobs/garden-cni/templates/pre-start.erb) it and set proper permissions
or
- Make garden-external-networker create the directory with permissions 700 [instead of 600](https://github.com/cloudfoundry/cf-networking-release/blob/2.22.0/src/garden-external-networker/bindmount/mounter_linux.go#L15)
or
- BPM could ensure [not only directory owner](https://github.com/cloudfoundry/bpm-release/blob/v1.0.4/src/bpm/runc/adapter/adapter.go#L109) but also its permissions

**More generic fix**
- Every directory should have its owner and the owner should ensure  the directory existence and permissions (this is what package managers, such as rpm, dpkg, pacman, normally do)

**Comment by giner** [comment]
The quickest and simplest workaround I could come up with for now:
```
cd $(mktemp -d)

git clone https://github.com/cloudfoundry/cf-networking-release
cd cf-networking-release
git checkout 2.15.1
git submodule update --init --recursive

patch -p1 << 'EOF'
diff --git a/jobs/garden-cni/templates/pre-start.erb b/jobs/garden-cni/templates/pre-start.erb
index e5cffcf7..c39b019b 100644
--- a/jobs/garden-cni/templates/pre-start.erb
+++ b/jobs/garden-cni/templates/pre-start.erb
@@ -1,3 +1,4 @@
 #!/bin/bash -eu

 rm -rf /var/vcap/data/garden-cni || true
+mkdir -m 700 /var/vcap/data/garden-cni
EOF

bosh create-release --name=cf-networking-fix --version=2.15.1-race-fix --force
bosh upload-release --name=cf-networking-fix
```
Though it will require recreating the release manully every time we bump cf-networking.

**Comment by KauzClay** [comment]
Hi @giner ! Thanks for such a thorough issue description, it really helped. We opted to merge in your PR #67.
